auth/oidc/identityprovider.go

```diff
@@ -261,27 +274,101 @@ func (p *Provider) ExtractIdentity(route *auth.RouteInfo, w http.ResponseWriter,
 	}
 
 	// Extract custom claims
-	var claims struct {
-		Name       string `json:"name"`
-		GivenName  string `json:"given_name"`
-		FamilyName string `json:"family_name"`
-		Email      string `json:"email"`
-		Verified   bool   `json:"email_verified"`
-	}
+	var claims claimsData
 	if err := idToken.Claims(&claims); err != nil {
 		log.Log(ctx, errors.Wrap(err, "parse claims"))
 		return nil, auth.Error(fmt.Sprintf("Invalid response from %s server.", name))
 	}
+
 	if claims.Name == "" {
 		// We *should* always get name with the profile scope, but fall back to joining the given and family names
 		// for misbehaving servers.
 		claims.Name = strings.TrimSpace(claims.GivenName + " " + claims.FamilyName)
 	}
 
+	if claims.Email == "" || claims.Name == "" {
+		provider, err := p.provider(ctx)
+		if err != nil {
+			log.Log(ctx, errors.Wrap(err, "retrieving OIDC provider from config"))
+			return nil, auth.Error("Cannot retrieve OIDC provider from config")
+		}
+
+		info, err := provider.UserInfo(ctx, oaCfg.TokenSource(ctx, oauth2Token))
+		if err != nil {
+			log.Log(ctx, errors.Wrapf(err, "fetching userinfo for user %s", idToken.Subject))
+			return nil, auth.Error("Cannot fetch OIDC user info")
+		}
+
+		if err := p.fetchUserInfo(ctx, &claims, info); err != nil {
+			return nil, err
+		}
```

Comment:
Currently, if you don't have email address, auth fails. This patch doesn't aim to add more functionality: it merely extends possibilities to collect full name and email address. Then, it's a good next step not to fail at this place, but to offer completing the profile instead. I'd favor doing this separately.

Reply:
Not all auth servers are equal: `email` scope can be consented/supported and still have `email` be empty (i.e. not all users in AD have a company email). In that case, the user can still login/create an account today as far as GoAlert will allow, so long as we get the token from the OIDC server it will still work (same with GitHub/GHE).

New users are only required to have a name and subject ID to login, that's handled in the user Store validation. Existing users *only* need a matching subject ID, and can log in with a blank name since it goes unused in that case. We **do** need to make sure that doesn't break -- otherwise, it leaves a potential that updating GoAlert, even with no other changes, would cause a user to be unable to login that could have before.

Likewise, setting `UserInfoEmailPath` shouldn't add any possibility to fail login that previously would have succeeded; rather it should only augment missing data where possible.

In that case, we still identified the user, they entered their user/pass to login at OIDC, and we got a valid token/subject ID.

Originally email was a strict requirement for new users but was relaxed early on since there was no actual need for it other than informational.

Existing-user login never had any data requirements, however, which is the bigger issue as a failure here could prevent login unnecessarily for existing users.